walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,84 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.io.*; 
+import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Arrays;
+
 
 public final class FindMeetingQuery {
+
+  /*
+  * Receive events and a request and returns all the available time intervals
+  */  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      Collection<TimeRange> result;
+      Collection<String> requestAttendees = request.getAttendees();
+      Collection<String> requestOptionalAttendees = request.getOptionalAttendees();
+      
+      ArrayList<TimeRange> resultList = new ArrayList<TimeRange>();
+      ArrayList<TimeRange> resultWithOptional = new ArrayList<TimeRange>();
+      
+      int previousEventEndTime = TimeRange.START_OF_DAY;
+      int previousEventTimeOptionalAttendees = TimeRange.START_OF_DAY;
+      long requestedTime = request.getDuration();
+
+      result = Arrays.asList();
+      
+      if(TimeRange.END_OF_DAY < requestedTime){
+            return result;
+      }
+
+      if(events.size() == 0){
+          result = Arrays.asList(TimeRange.WHOLE_DAY);
+          return result;
+      }
+      
+
+     for (Event event : events) {
+         Collection<String> eventAttendees = event.getAttendees();
+         TimeRange eventInterval = event.getWhen();
+         int eventStartTime = eventInterval.start();
+         if(requestAttendees.containsAll(eventAttendees)){
+            if(previousEventEndTime + requestedTime <= eventStartTime){
+                    resultList.add(TimeRange.fromStartEnd(previousEventEndTime, eventStartTime, false));
+            }
+            if(previousEventEndTime <  eventInterval.end())
+                previousEventEndTime =  eventInterval.end();
+
+            if(previousEventTimeOptionalAttendees + requestedTime <= eventStartTime){
+                resultWithOptional.add(TimeRange.fromStartEnd(previousEventTimeOptionalAttendees, eventStartTime, false));
+                }
+            if(previousEventTimeOptionalAttendees <  eventInterval.end())
+                previousEventTimeOptionalAttendees =  eventInterval.end();
+            } 
+
+         if(requestOptionalAttendees.containsAll(eventAttendees)){
+            if(previousEventTimeOptionalAttendees + requestedTime <= eventStartTime){
+                    resultWithOptional.add(TimeRange.fromStartEnd(previousEventTimeOptionalAttendees, eventStartTime, false));
+                }
+            if(previousEventTimeOptionalAttendees <  eventInterval.end())
+                    previousEventTimeOptionalAttendees =  eventInterval.end();      
+            }
+        }
```

Comment:
This for loop is too large and has a lot of conditions. Break it down in multiple functions whose names specify what are they doing. In the ideal case, the code itself should be the documentation of what it's doing, if you can't tell from a glance it's probably because you're missing some modularization.

Reply:
Solved